ibis/expr/scope.py

```diff
@@ -41,19 +41,72 @@
 from ibis.expr.timecontext import TimeContextRelation, compare_timecontext
 from ibis.expr.typing import TimeContext
 
-ScopeItem = namedtuple('item', ['value', 'timecontext'])
+ScopeItem = namedtuple('item', ['timecontext', 'value'])
 
 
 class Scope:
-    def __init__(self, items: Dict[str, ScopeItem] = None):
-        self._items = items or {}
+    def __init__(
+        self,
+        param: Dict[Node, Any] = None,
+        timecontext: Optional[TimeContext] = None,
+    ):
+        """ Take a dict of `op`, `result`, create a new scope and save
+        those pairs in scope. Associate None as timecontext by default.
+        This is mostly used to init a scope with a set of given params.
+        """
+        self._items = (
+            {op: ScopeItem(timecontext, value) for op, value in param.items()}
+            if param
+            else {}
+        )
 
     def __contains__(self, op):
+        """ Given an `op`, return if `op` is present in Scope.
+        Note that this `__contain__` method doesn't take `timecontext`
+        as a parameter. This could be used to iterate all keys in
+        current scope, or any case that doesn't care about value, just
+        simply test if `op` is in scope or not.
+        When trying to get value in scope, use `get_value(op, timecontext)`
+        instead. Because the cached data could be trusted only if:
+        1. `op` is in `scope`, and,
+        2. The `timecontext` associated with `op` is a time context equal
+           to, or larger than the current time context.
+        """
         return op in self._items
 
     def __iter__(self):
         return iter(self._items.keys())
 
+    def set_value(
+        self, op: Node, timecontext: Optional[TimeContext], value: Any
+    ):
```

Comment:
can you type `-> None` as the return (I assume it doesnt return anything)

Reply:
Yes I added type to the return value.